tests/test_ghostlastmatrixadapter.cpp

```diff
@@ -0,0 +1,331 @@
+  the Free Software Foundation, either version 3 of the License, or
+  (at your option) any later version.
+
+  OPM is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with OPM.  If not, see <http://www.gnu.org/licenses/>.
+*/
+#include <config.h>
+
+#define BOOST_TEST_MODULE PreconditionerTest
+#define BOOST_TEST_NO_MAIN
+#include <boost/test/unit_test.hpp>
+#include <opm/simulators/linalg/PreconditionerFactory.hpp>
+#include <opm/simulators/linalg/PropertyTree.hpp>
+#include <opm/simulators/linalg/getQuasiImpesWeights.hpp>
+#include <opm/simulators/linalg/WellOperators.hpp>
+
+#include <dune/common/parallel/mpihelper.hh>
+#include <dune/common/fmatrix.hh>
+#include <dune/common/unused.hh>
+#include <dune/common/parallel/indexset.hh>
+#include <dune/common/parallel/plocalindex.hh>
+#include <dune/common/parallel/communication.hh>
+#include <dune/istl/bcrsmatrix.hh>
+#include <dune/istl/owneroverlapcopy.hh>
+#include <dune/istl/schwarz.hh>
+
+class MPIError {
+public:
+    /** @brief Constructor. */
+    MPIError(std::string s, int e) : errorstring(s), errorcode(e){}
+    /** @brief The error string. */
+    std::string errorstring;
+    /** @brief The mpi error code. */
+    int errorcode;
+};
+
+void MPI_err_handler(MPI_Comm *, int *err_code, ...){
+    char *err_string=new char[MPI_MAX_ERROR_STRING];
+    int err_length;
+    MPI_Error_string(*err_code, err_string, &err_length);
+    std::string s(err_string, err_length);
+    std::cerr << "An MPI Error ocurred:"<<std::endl<<s<<std::endl;
+    delete[] err_string;
+    throw MPIError(s, *err_code);
+}
+
+typedef Dune::OwnerOverlapCopyAttributeSet GridAttributes;
+typedef GridAttributes::AttributeSet GridFlag;
+typedef Dune::ParallelLocalIndex<GridFlag> LocalIndex;
+
+template<class M, class G, class L, int n>
+void setupPattern(int N, M& mat, Dune::ParallelIndexSet<G,L,n>& indices, int overlapStart, int overlapEnd,
+                  int start, int end);
+
+template<class M>
+void fillValues(int N, M& mat, int overlapStart, int overlapEnd, int start, int end);
+
+
+template<class M, class G, class L, class C, int n>
+M setupAnisotropic2d(int N, Dune::ParallelIndexSet<G,L,n>& indices,
+                     const C& p, int *nout, typename M::block_type::value_type eps=1.0);
+
+
+template<class M, class G, class L, int s>
+void setupPattern(int N, M& mat, Dune::ParallelIndexSet<G,L,s>& indices, int overlapStart, int overlapEnd,
+                  int start, int end)
+{
+    int n = overlapEnd - overlapStart;
+
+    typename M::CreateIterator iter = mat.createbegin();
+    indices.beginResize();
+
+    for(int j=0; j < N; j++)
+        for(int i=overlapStart; i < overlapEnd; i++, ++iter) {
```

Comment:
Don't we need to change this,  to make the sure that the guys that are `GridAttributes::copy` come after the rest?

I think we would need some kind of mapping of the indices to ensure that?

Not sure whether we want an expensive check in GhostLastMatrxAdapter that checks the `GridAttributes::owner` comes first.

Reply:
Can we discuss this tomorrow once more and then eventually merge?